lms/services/digest.py

```diff
@@ -189,128 +178,90 @@ def unified_courses(self):
-
-        In the case of a course group this will return the given
-        authority_provided_id itself.
-
-        In the case of a sub-group this will return the authority_provided_id
-        of the course group that the sub-group belongs to.
-        """
-        first_grouping = self._db.scalars(
-            select(Grouping).filter_by(authority_provided_id=authority_provided_id)
-        ).first()
-
-        if not first_grouping:
-            return None
-
-        if first_grouping.type == Grouping.Type.COURSE:
-            return first_grouping.authority_provided_id
-
-        assert first_grouping.parent.type == Grouping.Type.COURSE
-        return first_grouping.parent.authority_provided_id
-
-    def _course_groupings(self, authority_provided_id):
-        """
-        Return all course groupings with the given authority_provided_id.
-
-        When an LMS has multiple application instances this may return multiple
-        course groupings with the same authority_provided_id but different
-        application instances.
-        """
-        return self._db.scalars(
-            select(Course).filter_by(authority_provided_id=authority_provided_id)
-            # Sort by updated so that we use the most recently updated courses
-            # first when picking a course title etc.
-            .order_by(Course.updated.desc())
-        ).all()
-
-    def _sub_groupings(self, course_groupings):
-        """Return all sub-groupings of the given course groupings."""
-        return self._db.scalars(
-            select(Grouping).filter(
-                Grouping.parent_id.in_([course.id for course in course_groupings])
-            )
-        ).all()
-
-    def _instructor_h_userids(self, courses):
-        """Return the h_userids of all instructors in the given courses."""
-        return tuple(
-            unified_user.h_userid
-            for unified_user in self.unified_users.values()
-            if self._is_instructor(unified_user, courses)
-        )
-
-    def _is_instructor(
-        self, unified_user: UnifiedUser, courses: Iterable[Course]
-    ) -> bool:
-        """Return True if `user` is an instructor in any of `courses`."""
-        return bool(
-            self._db.scalar(
-                select(func.count())
-                .select_from(AssignmentMembership)
-                .join(Assignment)
-                .join(AssignmentGrouping)
-                .join(LTIRole)
-                .filter(
-                    AssignmentGrouping.grouping_id.in_(
-                        (course.id for course in courses)
-                    )
-                )
-                .filter(AssignmentMembership.user_id.in_(unified_user.user_ids))
-                .filter(LTIRole.type == "instructor", LTIRole.scope == "course")
-            )
+    @property
+    def _h_userids(self):
+        return set(
+            self._audience
+            + [annotation["author"]["userid"] for annotation in self._annotations]
```

Comment:
This property is actually only used in one place so it could just be inlined but I thought it was more readable to separate out this little bit of logic

Reply:
Yeah I think you're right, great spot :+1: `context.unified_users` is now only used in one place which is to look up the names and email addresses of the audience users:

https://github.com/hypothesis/lms/blob/1b981bac0af2724a447aa8b162892ff2d83a051e/lms/services/digest.py#L50

So this no longer needs to contain the annotators :+1:

Done: https://github.com/hypothesis/lms/pull/5234/commits/db6178d12b973f0a53773aea8e7f7fa220adbd07